libs/unity/library/Runtime/Scripts/Media/MediaLine.cs

```diff
@@ -141,36 +156,48 @@ public MediaTrack SenderTrack
         /// If <see cref="Transceiver"/> is valid, that is a first session negotiation has already been conducted,
         /// then changing this value raises a <see xref="WebRTC.PeerConnection.RenegotiationNeeded"/> event on the
         /// peer connection of <see cref="Transceiver"/>.
-        /// </summary>
+        ///
+        /// Must be changed on the main thread.
+        /// </remarks>
         public MediaReceiver Receiver
         {
             get { return _receiver; }
             set
             {
-                if (value == null)
+                if (_receiver == value)
                 {
-                    if (_receiver != null)
-                    {
-                        _receiver.OnRemovedFromMediaLine(this);
-                        _receiver = null;
-                    }
+                    return;
                 }
-                else
+                if (value!= null && value.MediaKind != MediaKind)
                 {
-                    if (value.MediaKind != MediaKind)
+                    throw new ArgumentException("Wrong media kind", nameof(Receiver));
+                }
+
+                // If the connection is active, ensure that this is run on the main thread
+                // and doesn't race with the signaling handlers.
+                // If the connection is not awake we cannot use EnsureIsMainAppThread, but
+                // there should be no race on shared state, so skip the check.
+                if (_peer.isActiveAndEnabled)
+                {
+                    _peer.EnsureIsMainAppThread();
+                }
+
+                if (_receiver != null)
+                {
+                    if (_remoteTrack != null)
                     {
-                        throw new ArgumentException("Wrong media kind", nameof(Receiver));
+                        _receiver.OnUnpaired(_remoteTrack);
                     }
-                    if (_receiver != value)
+                    _receiver.OnRemovedFromMediaLine(this);
+                }
+                _receiver = value;
+                if (_receiver != null)
+                {
+                    if (_remoteTrack != null)
                     {
-                        if (_receiver != null)
-                        {
-                            _receiver.OnRemovedFromMediaLine(this);
-                        }
-
-                        _receiver = value;
-                        _receiver.OnAddedToMediaLine(this);
+                        _receiver.OnPaired(_remoteTrack);
                     }
+                    _receiver.OnAddedToMediaLine(this);
```

Comment:
nit: reverse order for consistency with the unpair/remove above?
```
_receiver.OnAddedToMediaLine(this);
if (_remoteTrack != null) {
    _receiver.OnPaired(_remoteTrack);
}
```

Reply:
Fixed